File: hw/hw1/question2.jl
using Pkg;
Pkg.activate("../..");
using Flux, Random, Plots

"""
    mse(ŷ, y)

Mean squared error loss function
"""
mse(ŷ, y) = sum((ŷ - y) .^ 2) / length(y)

loss(model, x, y) = Flux.crossentropy(model(x), Flux.onehotbatch(y, 0:9))

function high_freq(x)
    return sin(5x) + cos(5x)
end

function low_freq(x)
    return sin(x)
end

function cantor(x, level = 10)
    if level == 0
        return x < 1 // 3 ? 2x : 2x - 1
    elseif x < 1 // 3
        return 0.5cantor(3x, level - 1)
    elseif x < 2 // 3
        return 0.5
    else
        return 0.5 + 0.5cantor(3x - 2, level - 1)
    end
end

function weierstrass(x, a = 0.5, b = 3, n = 25)
    return sum([a^n * cos(b^n * pi * x) for n = 0:n])
end

function periodic_data(x)
    return sin(x)
end

function nonperiodic_data(x)
    return x^2
end

function smooth_data(x)
    return sin(x) + cos(x)
end

function non_smooth_data(x)
    #sawtooth wave
    return x - floor(x)
end

function noisy_data(x)
    return sin(x) + 0.1 * randn()
end

function noiseless_data(x)
    return sin(x)
end

function smooth_vs_non_smooth(max_x, optimizer, num_epochs::Integer)

    func_type = "smooth vs. non-smooth"

    model1 = Chain(Dense(1 => 10, relu), Dense(10 => 1, sigmoid), softmax)
    model2 = Chain(Dense(1 => 10, relu), Dense(10 => 1, sigmoid), softmax)

    op_state_s = Flux.setup(optimizer, model1)
    op_state_ns = Flux.setup(optimizer, model2)

    x_train = rand(1000) .* max_x

    y_s_train = smooth_data.(x_train)
    y_ns_train = non_smooth_data.(x_train)

    data_s = [(x, y) for (x, y) in zip(x_train, y_s_train)]
    data_ns = [(x, y) for (x, y) in zip(x_train, y_ns_train)]


    convergence_smooth = zeros(num_epochs)
    convergence_non_smooth = zeros(num_epochs)
    for ii = 1:num_epochs
        println(ii)
        Flux.train!(model1, data_s, op_state_s) do m, x, y
            loss(m(x), y)
        end

        Flux.train!(model2, data_ns, op_state_ns) do m, x, y
            loss(m(x), y)
        end
    end

    # plot convergence_smooth and convergence_non_smooth in the same plot, label them and save it
    plot(
        convergence_smooth,
        label = "smooth",
        xlabel = "epochs",
        ylabel = "loss",
        title = "smooth vs. non-smooth",
    )
    plot!(convergence_non_smooth, label = "non-smooth")
    savefig("plots/smooth_vs_non_smooth.png")
end



function main()
    # investigate types of target function on fixed neural network on number of epochs
    # smooth vs non-smooth
    # Initialize the ADAM optimizer with default settings
    smooth_vs_non_smooth(5.0, Adam(), 500)
end


main()

File: hw/hw1/question2.py
import numpy as np
import matplotlib.pyplot as plt
import torch
from torch import nn

class NeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(1, 512),
            nn.ReLU(),
            nn.Linear(512, 1),
        )

    def forward(self, x):
        x = self.linear_relu_stack(x)
        return x


def high_freq(x):
    return np.sin(5*x) + np.cos(5*x)

def low_freq(x):
    return np.sin(x) + np.cos(x)

def cantor(x, level=10):
    if level == 0:
        return 2*x if (x < 1 / 3) else (2*x -1)
    elif x < 1 / 3:
        return cantor(3*x, level-1) / 2
    elif  x < 2 / 3:
        return 1 / 2
    else:
        return cantor(3*x - 2, level-1) / 2 + 1 / 2


def weierstrass(x , a = 0.5, b = 3, n =25):
    return sum([a**i * np.cos(b**i * np.pi * x) for i in range(n)])

def periodic_data(x):
    return np.sin(2 * np.pi * x)

def non_periodic_data(x):
    return x**2

def smooth_data(x):
    return x

def non_smooth_data(x):
    #sawtooth wave
    return x - np.floor(x)

def noisy_data(x):
    return np.sin(2 * np.pi * x) + np.random.normal(0, 0.1, len(x))

def non_noisy_data(x):
    return np.sin(2 * np.pi * x)


def train_loop(x,y,model,num_epochs,learning_rate,loss_fn,optimizer):
    losses = []

    x = torch.from_numpy(x).view(-1,1).float()
    y = torch.from_numpy(y).view(-1,1).float()

    for epoch in range(num_epochs):
        # Compute prediction and loss
        pred = model(x)
        loss = loss_fn(pred, y)

        # Backpropagation
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()

        losses.append(loss.item())

    return losses



def main():

    model1 = NeuralNetwork()
    model2 = NeuralNetwork()

    num_epochs = 1000
    learning_rate = 0.001

    func_type = "werid functions"
    xtrain = np.random.uniform(0, 1, 200)
    ytrain_s = np.array([cantor(x) for x in xtrain])
    ytrain_ns = np.array([weierstrass(x) for x in xtrain])

    loss_fn = nn.MSELoss()
    optimizer1 = torch.optim.SGD(model1.parameters(), lr=learning_rate)
    optimizer2 = torch.optim.SGD(model2.parameters(), lr=learning_rate)

    convergence_s = train_loop(xtrain, ytrain_s, model1, num_epochs, learning_rate, loss_fn, optimizer1)
    convergence_ns = train_loop(xtrain, ytrain_ns, model2, num_epochs, learning_rate, loss_fn, optimizer2)

    plt.plot(convergence_s, label = "cantor")
    plt.plot(convergence_ns, label = "weierstrass")
    plt.legend()
    plt.title("Convergence of Neural Network for " + func_type)
    plt.xlabel("Epoch")
    plt.ylabel("Loss")
    plt.savefig(f"plots/convergence_{func_type}.png")
    plt.show()



if __name__ == '__main__':
    main()
